Setup mysql script for CentOS 7

kimlew opened the pull request:
@kcharter 
WRT: `setup_mysql.sh`
- to start MySQL on a CentOS 7 VM

Can you please look at what I have so far. In particular, I am having problems with: 

1. re-creating the situation that caused an error to add the code starting at line 178:
`if ! MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}'); then ...`
I can't re-create the situation regardless of the combination of removing MySQL client, MySQL server, the data files or the log file, so that part of the code is run. 

2. making sure I test all the combinations of deleting those 4 things and put in the checks I need. 

Any suggestions?

kimlew committed:
Adds .flaskenv and .vagrant/

kimlew committed:
Adds directive to create centos/7 box

kimlew committed:
Creates file to use with scp -F from vagrant ssh-config

kimlew committed:
Runs but doesn't accept root password

kimlew committed:
Rearranges code, creates function & reaches running mysql_secure_installation

kimlew committed:
Responds to diff expect output. Checks MySQL install & mysql_secure_installation run

kimlew committed:
Changes if to not use [command] so Bash set -e doesn't exit you from script

kimlew committed:
Moves temp password so global again & adds missing end curly brace

kimlew committed:
Adds extra ( which seems to be their typo

kimlew committed:
Adjusts check for MySQL package repo & MySQL server, Adds sudo to rpm command

kimlew committed:
Adds check for when /var/log/mysqld.log exists but the temp password is gone

kimlew committed:
Adds important step of creating mysql_pw.txt

kimlew committed:
Uncomments line, vb.gui = true - to solve error w signal 10

kimlew committed:
Adds TODO to run setup_mysql.sh to install MySQL

kimlew committed:
Initial commit of Python Flask app folders and files

kcharter commented on setup_mysql.sh:
```diff
@@ -0,0 +1,230 @@
+  echo "new root password: ${MYSQL_ROOT_PWD}"
+  echo
+
+SECURE_MYSQL=$(expect -f '-' <<HERE
```
You don't need to capture the output of `expect` in the `SECURE_MYSQL` variable and then `echo "$SECURE_MYSQL"` below. Instead you can just run `expect` directly. Its standard output and standard error will still appear on the terminal.

So instead of

```sh
SECURE_MYSQL=$(expect -f '-' <<HERE
...
HERE
)
echo "$SECURE_MSQL"
```

you can simplify it to

```sh
expect -f '-' <<HERE
...
HERE
```

kimlew replied:
Removed unneeded code & re-tested.

kcharter commented on setup_mysql.sh:
```diff
@@ -0,0 +1,230 @@
+sudo systemctl start mysqld.service
+sudo systemctl status mysqld
+
+# CHECK CASE: At this point, MySQL is installed, mysql_secure_installation has
+# NOT been run, BUT mysqld.log file has lost temporary password.
+# Note: Without the if, when this line is run and fails, you are exited from script.
+# MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}')
```
You might have run `mysql_secure_installation` on line 122 above, so this comment might be wrong.

However, your script might be a little simpler if you _don't_ run `mysql_secure_installation` on line 122.

kimlew replied:
I think you meant for me to fix the comment, which I have.
Then the call for the function, `run_mysql_secure_installation` doesn't occur till about Line 209 - which I think, is what you meant about NOT running `mysql_secure_installation` earlier on Line 122, right?

kcharter replied:
Yes, that's right.

kcharter commented on setup_mysql.sh:
```diff
@@ -0,0 +1,230 @@
+# Note: Without the if, when this line is run and fails, you are exited from script.
+# MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}')
+
+if ! MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}'); then
```
Ah, you were saying you can never get this to fail -- I think you said you even tried deleting the `/var/log/mysqld.log` file.

One reason you might not be able to get it to fail could be that we need to put the command

```sh
set -o pipefail
```

at the top of the script where you run

```sh
set -e
```

This changes the shell so that in a command pipeline like

```sh
sudo grep ... | awk ...
```

if the first command exits with a non-zero status code, then the whole pipeline exits with that status code. By default, the pipeline will exit with the status code of the last command. I forgot to tell you about `set -o pipefail`, so the problems you are having here are probably my fault! Sorry!

kimlew replied:
Combined line to: `set -eo pipefail` & re-tested.

kcharter commented on setup_mysql.sh:
```diff
@@ -0,0 +1,230 @@
+  echo "MySQL package repos AND MySQL server are already installed."
+  echo
+
+  # Case 1: CHECK if assignedroot password works with MySQL.
+  # If it root password exists, exit with 0 & skip rest of code since it means
+  # MySQL is installed AND mysql_secure_installation has already been run.
+  # Case 2: else - Only the temporary password exists so far which means MySQL
+  # is installed BUT mysql_secure_installation has NOT been run yet, so run it.
+  echo "TEST if root password runs MySQL, i.e., mysql_secure_installation was already run."
+  echo
+  if expect -f '-' <<HERE
+  set timeout 5
+  spawn mysql -u root -p --execute "quit"
+  expect "Enter password:"
+  send "$MYSQL_ROOT_PWD\r"
+  expect eof
+  exit [lindex [wait] 3]
+HERE
+  then
+    echo
+    echo "ROOT password assigned in mysql_pw.txt WORKED! MySQL is installed & mysql_secure_installation was run."
+    exit
+  else
+    echo
+    echo "NO assigned ROOT password yet. RUNNING mysql_secure_installation..."
+    echo
+    run_mysql_secure_installation
+  fi
```
I think you don't need this code from lines 99 through 123, because you already handle running `mysql_secure_installation` later in the script.

kimlew replied:
I think that code is still a valuable check to exit early if an assigned root password exists, i.e., MYSQL_ROOT_PWD. So I moved it to after Line 146-147:  
```
sudo systemctl start mysqld.service
sudo systemctl status mysqld
```
So it is now before Line 169:  
`# CHECK ODD CASE`
and before Line 188:    
`# CHECK if temporary password can be used to RUN MySQL.`

And I removed the `else` chunk.

**Bonus**: This change also took care of an error I saw before:
...
spawn mysql_secure_installation

Securing the MySQL server deployment.

Enter password for user root:
Error: Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)
expect: spawn id exp6 not open
    while executing
"expect "New password:""

kimlew replied:
@kcharter Can you have another look at this PR?
I created a new clean VM, copied over the required files & re-tested the script in various states of removal with these commands:  
```
sudo yum erase mysql  # Removes the client.
sudo rm -rf /var/lib/mysql  # Deletes the database files.*
sudo yum erase mysql-community-server  # Deletes server config files.
sudo rm /var/log/mysqld.log
```
It works the same way with expected outcomes, as it did on the old VM. I didn't find any new error messages.

kcharter replied:
Yep, looking at it this afternoon.

kcharter reviewed (commented):
@kimlew I have a couple of comments and suggestions.  One of them is an oversight I made that might help with that code you expected to run when the temporary password wasn't in `/var/log/mysqld.log`.

Try the suggestions out and I will review again.

kimlew committed:
Adds setup_mysql.sh to steps

kimlew committed:
Removes declaration of SECURE_MYSQL... & echo "SECURE_MYSQL"

kimlew committed:
Edits some comments

kimlew committed:
Adds set -o pipefail to exit when 1st command in line with pipe fails

kimlew committed:
Corrects typo from even to echo & fixes comment

kimlew committed:
Moves check to run mysql with MYSQL_ROOT_PWD & edits comments

- moves check before ODD CASE & running mysql_secure_installation

kimlew committed:
Re-orders MySQL steps

kimlew committed:
Edits comments so clearer in setup_machine.sh & adds VM name to Vagrantfile

kimlew committed:
Corrects wrong location of comment

kcharter commented on README.md:
```diff
@@ -1,26 +1,34 @@
   `vagrant ssh-config > vagrant-ssh-config`
 
-3. Make sure you created mysql_pw.txt outside of this script & put in the password you want for root for login to MySQL. This script assumes mysql_pw.txt already exists.
+3. Copy `setup_machine.sh` & `setup_mysql.sh` with `scp` to the virtual machine/deployment machine, e.g.,<br>
+  `scp -F vagrant-ssh-config setup_machine.sh vagrant@default:setup_machine.sh`<br>
+  `scp -F vagrant-ssh-config setup_mysql.sh vagrant@default:setup_mysql.sh`
```
If you want those lines indented in a code block I believe you can do it without the `<br>` tags by using triple back-quotes _and_ indenting the code block at least four spaces. I think this works:

````
2. Copy `setup_machines.sh` & `setup_mysql.sh` with `scp` to the virtual machine/deployment machine
    ```
    scp -F vagrant-ssh-config setup_machine.sh vagrant@default:setup_machine.sh
    scp -F vagrant-ssh-config setup_mysql.sh vagrant@default:setup_mysql.sh
    ```
````

kimlew replied:
Done. That is better with 4 spaces to indent. Atom changed the 2 spaces I added to the end of lines for a new line to those <br> tags.

kcharter reviewed (commented):
This looks good. I have one little suggestion about your `README.md` and I would like to try running your script myself.

kimlew committed:
Replaces Atom-inserted <br> tags with 4 spaces to indent & -x flags

kimlew committed:
Adds cd .. pipenv run flask run & port 8084 parts, etc.

kimlew committed:
Forwards port with guest: 5000, host: 8084, host_ip: "127.0.0.1"

kimlew committed:
Adds part that runs MySQL with root password & creates database & table

kimlew committed:
Creates .env with MySQL host, user, password, db

kimlew committed:
Updates some versions

kimlew committed:
Adds load_dotenv() which is needed to use .env

kimlew committed:
Copies project folder, setup_mysql.sh, setup_machine.sh to Vagrant CentOS 7 VM

kimlew committed:
Adds .gitignore

kimlew commented:
@kcharter Can you look at the changed files? Thanks.

kcharter commented on copy_req_files_to_vm.sh:
```diff
@@ -0,0 +1,65 @@
+
+# --- Project files & folders for app, pythonapp-createjazzlyric ---
+# Note: scp -r  recursively copies entire directory, pythonapp-createjazzlyric
+# which includes:
+# README.md
+# conn_vars_dict.py
+# create_mysql_db.sql
+# Pipfile
+# Pipfile.lock
+# create_jazz_lyric.py
+# static
+# templates
```
I wouldn't bother listing all these files in the comment, because as the contents of `pythonapp-createjazzlyric` change you'll have to remember to update this comment. I think it's sufficient to just say that you're copying the directory and all its contents.

kimlew replied:
Good point about later maintenance of comment. Removed listing of files.

kcharter commented on setup_machine.sh:
```diff
@@ -40,9 +44,20 @@ cd pythonapp-createjazzlyric
+    echo "DB_PASSWORD=$(cat ../mysql_pw.txt)"
+    echo 'DB_NAME=lyric_db'
+  } > .env
+fi
```
Suggestion: if `../mysql_pw.txt` doesn't exist, you will probably want to fail the script with a message (I was being a bit sloppy when I suggested adding this part). You could do this by adding an `else` where you echo a message and invoke `exit 1`.

kimlew replied:
Good point about failing early if the required file, `mysql_pw.txt`, is missing. Added `else` clause.

kimlew replied:
Also, I added a check for if Python 3.9 is already installed, which you mentioned several lines above, so the 7 lines don't need to be re-run:
```
if [[ $(python3.9 --version) != "Python 3.9.7" ]]; then
  echo "Getting and installing Python 3.9.7..."
  wget https://www.python.org/ftp/python/3.9.7/Python-3.9.7.tgz
  ...
```

kcharter replied:
OK, you have addressed the issue.

kcharter commented on copy_req_files_to_vm.sh:
```diff
@@ -0,0 +1,65 @@
+DEST_PROJ_DIR='/home/vagrant/pythonapp-createjazzlyric/'  # On VM.
+
+# --- Files needed at root directory on VM to set up VM. ---
+scp -F "${VAG_SSH_CFG}" "${SRC_REQ_FILES_DIR}"setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"setup_mysql.sh
+scp -F "${VAG_SSH_CFG}" "${SRC_REQ_FILES_DIR}"setup_machine.sh vagrant@default:"${DEST_ROOT_DIR}"setup_machine.sh
```
I noticed several days ago when you were showing me this script that you included a trailing slash in the value of `SRC_REQ_FILES_DIR`. Did you do that just so you wouldn't have to insert a slash after the variable interpolation like this:

```
"${SRC_REQ_FILES_DIR}"/setup_mysql.sh
```

I ask because I usual dislike having to remember to put a trailing slash on the value of the variable, because it's not obvious from the variable name and it's easy to forget. But you might have a good reason for doing it that I'm not seeing.

kimlew replied:
I DID add the trailing / to the ends of the paths for the directory variables, because I always forget to manually add the trailing / in the actual command.
I also find it harder to read when it is all clumped together in the command. For example:

Harder to Read:
`scp -F "${VAG_SSH_CFG}"/"${SRC_REQ_FILES_DIR}"setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"setup_mysql.sh`

Easier to Read:
`scp -F "${VAG_SSH_CFG}" "${SRC_REQ_FILES_DIR}"setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"setup_mysql.sh`

There sure seems to be mixed opinions on this, without a conclusive answer on whether it is BAD Practice to have a trailing / in a directory variable name. If the Common/Best Practice for this is to show the / in the command, I can remove the / from the variable names.

Here are some opinions I found with a strong opinion from 1 person:

`https://stackoverflow.com/questions/980255/should-a-directory-path-variable-end-with-a-trailing-slash`
...
It is expected that a directory not end in a slash. Any expectation that a directory ends in a slash is wrong. Thus adding a slash to the end of a *_DIR variable's value would be subverting expectations.
...

But that is just 1 person's strong opinion. If you find documentation that supports this, I would be happy to change my code.

kcharter replied:
I think you can leave it as is as long as you are consistent about `_DIR` variables having values that end in slashes.

Here's my longer response.

You're right that there are mixed opinions on this, and partly it's because paths are a little quirky. Ordinarily the slash `/` is just a separator, except when it appears all by itself and means the root directory. And, as you have learned, different tools (like `rsync`) ascribe special meaning to a trailing slash, but most tools ascribe no special meaning to a trailing slash.

You said

> Harder to Read:
> scp -F "${VAG_SSH_CFG}"/"${SRC_REQ_FILES_DIR}"setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"setup_mysql.sh
> 
> Easier to Read:
> scp -F "${VAG_SSH_CFG}" "${SRC_REQ_FILES_DIR}"setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"setup_mysql.sh
> 

But did you mean for the "Harder to Read" example to be:

```
scp -F "${VAG_SSH_CFG}" "${SRC_REQ_FILES_DIR}"/setup_mysql.sh vagrant@default:"${DEST_ROOT_DIR}"/setup_mysql.sh
```

I ask because I was suggesting putting a slash `/` after the variable substitutions for `${SRC_REQ_FILES_DIR}` and `${DEST_ROOT_DIR}`, like above, and leaving it out of the values of the `_DIR` variables.

The reason I suggested this is I would personally find it hard to remember that if I change the values assigned to `SRC_REF_FILES_DIR` and `DEST_ROOT_DIR` that I have to add trailing slashes. The names of the variables don't suggest to me that they need a trailing slash, so I might forget. However, you personally can adopt the convention that a `..._DIR` variable always has a value that ends in a slash if you like. If you adopt that convention and use it consistently, then that's fine too.

One commenter pointed out that if you keep things the way you have them then you could use an empty value for `SRC_REF_FILES_DIR` to mean the present working directory. Personally I think that's a little fragile and it "works by accident". You can refer to the present working directory more directly with `${PWD}` and I would prefer that, but that's just me.

kcharter reviewed (commented):
I've got a couple of very minor comments to add to the new changes, nothing really complicated though.

One thing additional thing I'd like to suggest for `setup_machine.sh`: check if the version of Python you want is already installed and, if it is, don't bother downloading, building, and installing a new Python. This will make `setup_machine.sh` run quite a bit faster after the first time it runs on a VM.

kimlew committed:
Adds vagrant-ssh-config

kimlew committed:
Adds 2 checks, 1 for Python3.9 & 1 for mysql_pw.txt

kimlew committed:
Removes listing of project files from comments

kimlew committed:
Removes vagrant-ssh-config

kcharter commented on setup_machine.sh:
```diff
@@ -18,17 +18,21 @@ set -e
-sudo rm Python-3.9.7.tgz*
 
+# Check if Python aleady installed. If already there, skips next 7 lines.
+if [[ $(python3.9 --version) != "Python 3.9.7" ]]; then
```
You should quote the process substitution, i.e.

```
if [[ "$(python3.9 --version)" != "Python 3.9.7" ]]; then
```

The quotes guard against the possibility that `$(python3.9 --version)` produces nothing. Without the quotes, an empty result will cause a syntax error in your script, which would be confusing (`shellcheck` would have caught this I think)

kcharter replied:
Ah, I just learned something I didn't know. You are safe because you used the double-square-brackets `[[ ... ]]`.

I put this little test script in a file:

```
#!/bin/bash

if [[ $(python --version) != "Python 3.9.7" ]]; then
    echo nope
fi
```

and `shellcheck` doesn't complain about it. I ran it and there is no syntax error and it prints `nope` -- my Python is version 2.7.18 so the body of the `if` runs.

But when I change the `[[ ... ]]` to single brackets `[ ... ]`  like this:

```
#!/bin/bash

if [ $(python --version) != "Python 3.9.7" ]; then
    echo nope
fi
```

then `shellcheck` does complain:

```
MacBook-Air:tmp kcharter$ shellcheck junk.sh

In junk.sh line 3:
if [ $(python --version) != "Python 3.9.7" ]; then
     ^-----------------^ SC2046 (warning): Quote this to prevent word splitting.

For more information:
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
```

and if I try running the script it fails with a syntax error:

```
MacBook-Air:tmp kcharter$ ./junk.sh
Python 2.7.18
./junk.sh: line 3: [: !=: unary operator expected
```

kcharter replied:
To be on the safe side, you should redirect standard error to standard output when you run `python3 --version` like so:

```
$(python3 --version 2>&1)
```

The reason is that some versions of Python will print the version on standard error rather than standard output. The Python 2 on my Mac does this. The Python 3 I have (3.9.9) prints it on standard output so your script will probably work, but I think it's a good idea to include the redirection just to be safe.

I learned this by trying a little experiment with redirecting the standard output to `/dev/null`. For `python`, which is Python 2.7.18 on my machine, I still see the version when I redirect standard output to `/dev/null`:

```
MacBook-Air:tmp kcharter$ python --version > /dev/null
Python 2.7.18
MacBook-Air:tmp kcharter$
```

but for `python3` I don't:

```
MacBook-Air:tmp kcharter$ python3 --version > /dev/null
MacBook-Air:tmp kcharter$
```

This shows that `python --version` is printing to standard error, but `python3 --version` is printing to standard output. Maybe they changed this in Python 3 and all Python 3 versions will print the version on standard output, but it's hard to say for sure.

You can make sure you capture the version string no matter what by redirecting standard error to standard output with `2>&1`.

kimlew replied:
a. Every time I ran shellcheck, it never flagged needing double quotes. I'm pretty sure this is why almost everything I read advises to use `[[ ]]` around a comparison. I added `" "` around `"$(python3.9 --version)"` mostly since the line looks more syntactically correct/balanced in the comparison:
`if [[ "$(python3.9 --version)" != "Python 3.9.7" ]]; then`

b. Since I state in the comparison to check for the version, `python3.9`, I'm going to leave out your suggestion of adding in `2>&1`. From most of what I read, this is only an issue for Python versions Python 3.5 or earlier & since I was specific to a much later version, I won't have this issue.

kcharter reviewed (commented):
One additional comment. I'm still working on my response to your question about the trailing slashes.

kimlew committed:
Adds " " around command substitution tho unneeded with [[ ]]

kimlew committed:
Adds change to root directory & checks for nano & wget

kimlew committed:
Adds import for expired MySQL gpg key

kimlew committed:
Re-words 2 user messages about MySQL root password

kimlew committed:
Removes . which includes scp of hidden files

- do not need to do now, since .env & .flaskenv are created at end of setup_machine.sh

kimlew committed:
Corrects steps

kimlew commented:
@kcharter I've re-run the entire process on a clean VM and it works.

There seem to be quite a few manual steps to run, which you can see in `README.md` and seems a bit clunky to me. Any suggestions?

kcharter commented:
You could write a little script that ties everything together and tries to run all the steps in your   README file.

- You can use the 'read' command to read the desired database password from the user (I believe it has a command option for reading passwords where it won't echo the characters to the terminal)
- you can use 'ssh' to run individual commands on the vagrant  box
- You can use a command called 'nohup' on the vagrant box to run 'pipenv run flask run' in the background without quitting when 'ssh' disconnects

I can give more detailed suggestions shortly if you are interested

kimlew committed:
Stops displaying password, checks for mysql_pw.txt, edits user messages & comments

kimlew committed:
Creates 1 main script that calls other scripts to simplify process

kimlew committed:
Adds reminder of where you are in mid-process

kimlew committed:
No change to file, but new copy created during process

kimlew commented:
@kcharter 
I think this is more user friendly now.
The `README.md` is substantially shorter and less confusing with the addition of `main.sh` that calls the other scripts. 
The visual check of the password is now earlier in the process and only shows the 1st and last characters of the password vs. the whole password which was BAD practice.
Can you re-review? Thanks.

kcharter commented on setup_mysql.sh:
```diff
@@ -46,8 +30,6 @@ run_mysql_secure_installation() {
   # Use HERE document vs. as a straight multiline string - to avoid quoting issues
   echo
   echo "GETTING assigned root password from file..."
```
Do you still need this `echo` after removing the assignment to `MYSQL_ROOT_PWD` that used to be on the next line?

kimlew replied:
@kcharter 
I have several lines as checks/debugging throughout the scripts, because I still find it hard to see that things are running correctly with the barrage of output in the terminal as the scripts run.

But, I remove the lines you pointed out, which also shorten the code a bit, and edited Lines 201 & 202, which come before the call of `run_mysql_secure_installation`. Lines 201+ are now:
```
  echo "Temporary root password for MySQL exists, so you have NOT run mysql_secure_installation yet."
  echo "WILL NOW RUN run mysql_secure_installation & set password to use the one from ${MYSQL_PW_FILE}."
  echo
  run_mysql_secure_installation
```

kcharter reviewed (commented):
This looks good, I have just a question about an `echo` statement you might be able to move or remove.

kimlew committed:
Removes & edits user messages

kcharter reviewed (approved):
Looks good to me!
